### DISCUSSION TITLE
Keymap Layering/Shifting with Assigned Key (Like Hypershift in Synapse for Keypads)

### DISCUSSION THREAD

**Discussion by MicHaeL-MonStaR** [sezanzeb/input-remapper#688] [Ideas]
First of all, I think it's great this program exists, because it fills a hole that companies refuse to fill *coughrazercough*, or at least are slow to do so. - However, despite that this exists, it still might lack some key functionality that is really important to the potential of a keypad.
Though, this post is partially a question, because there might be a way. - But if not, it's a suggestion or request.

**The main question is:** Is it possible to assign one key to shift the entire layout (or selected keys) to a different layout or programming/mapping?
I've looked around but couldn't really find anything. But considering you can input a lot of stuff, there might be some way I don't know about.

**If you don't know what I mean:** It's basically turning one key on the programmed device into a modifier, which would turn the programming of the device to what would practically be a different profile, except it's within the same profile and just another "layer", kind of like what a common keyboard does with its modifiers.
This feature is called "Hypershift" in Synapse, but it was already a feature for Belkin's Nostromo SpeedPad (which then became Razer's Nostromo, turning into Tartarus and Orbweaver), except it had a more generic name like "shifting" or "toggling" the layers or profiles. - You could still toggle through at least three "layers/profiles" in Synapse, but "Hypershift" is very specifically to swiftly "shift" to another layer for all keys and back when you release the modifier-button.

**The reason this is useful** is because of course keypads have way fewer keys available, in this case about 20, which isn't very different from half the main part of a keyboard, but that can be a bit limiting. - The "layering" of keymaps helps at least doubling their functionality, and the advantage is that you don't even need to move your hand or fingers. - What's more is that you can "group" the types of functionality to these "layers", such as your normal navigational keys on the main "non-shifted" layer, but then put things like item-keys (such as the numbers) on the "(Hyper)shifted" layer, or whatever you like.

_I had been using the Belkin Nostromo SpeedPad n52 this way since 2004/2005, and because it wore out I had to move to Razer's Tartarus (Pro) in 2019/2020. But now that I've moved to Linux, and there's zero Linux-support for Synapse (although tech-support told me they're working on it, albeit without any time-frame), a lot the functionality of the device is gone as it was all powered by Synapse. - It works, but all I can do is straight-forward single-layered mapping, unless now I'll be told that you can program Input Remapper to do more than I realize of course. - It just slashes the versatility down critically for me, enough to feel handicapped, because I don't even know how to play games with a keyboard anymore, and I even used it outside of games for productivity. Not to mention that Synapse also switched profiles dynamically on-the-fly. - So I'm basically back two decades in terms of interfacing with my computer, unless I go back to Windows._

_So any suggestions or even new features for this would be greatly appreciated, because I can't wait around for Razer to do anything about their software, even the Windows-version of it. - Thanks for reading, the consideration, and this program so far._

**Comment by SamuelGogarty** [comment]
I am also looking to do this.
@MicHaeL-MonStaR if you really need this functionality now, you can use the newest fork of keyboarding-master
https://github.com/mhackett909/Keyboarding-Re-Master
however in my opinion input-remapper is better, so I would really like if this feature was implemented, because I also don't see any straight-forward way to achieve this.

**Comment by jonasBoss** [comment]
Well there is no specific feature for that. But I am wondering do we need that? I think this is already possible. You can just create multiple mappings for all the keys you want to change in the second layer:

 - `Shift R + KP End` -> `a`
 - `Shift R + KP Down` -> `b`
 - `Shift R + KP Next` -> `c`
 - `Shift R + KP Left` -> `d`

this should behave the same as having right shift switch to a different "layer" or whatever you want to call it. Just add a mapping for each key you want to have different from the normal layout.

**Reply by jonasBoss** [reply]
you can even add more "layers" by using other keys or combination.

 - "layer" 1: `Shift R + other key`
 - "layer" 2: `Shift R + Alt R + other key`

**Reply by MicHaeL-MonStaR** [reply]
Sorry, it's not entirely clear to me the method you're describing. I'm really quite lost, because I don't understand what is what exactly, and what goes where.

What exactly does `Shift R + KP End -> a` for example mean? - Is that "Right Shift" and "Keypad End" programmed to a key, and then when you hit that key it becomes 'a'?

It seems to me that perhaps you might be looking at it more in the context of keyboards and key-combinations and macros(?). - Cause my intention might be way simpler function-switching for singular keys on keypads (particularly for their limited size).

Just to clarify what I mean to do, it's as follows: I'll take the classic example of using the big (non-descript) button on the side of a SpeedPad/Tartarus...
Say "Key 01" (they're numbered on this keypad, it could be a letter on a regular keyboard, but I don't want to make it confusing) has the letter A mapped to it, but when I press/hold the button on the side, the mapping of Key 01 will change to B for instance (and then when the button is pressed again or let go it changes back, I mostly use the latter method for speed).
This could be done for all keys at the same time, for as far as you change anything in that other "layer", otherwise any unchanged keys (or their mapping) will stay as what it was on the original/first "layer".

Is that what can be done with what you described?

Cause it's practically and potentially (but not necessarily) changing the entire profile, like choosing a language-layout in an OS, but with a key. - And, as an idea, that could actually be a simpler way to implement it, by just allowing to switch profiles (or presets in this case) with the press of a designated key on-the-fly. - There could be a feature programmed to be a "preset switch" on a key, and then choose which profile from a drop-down or something. - This would be a workaround to not have to program in "layering" within each preset and for each key. - It would immediately at least double the functionality, even for a keyboard.

However, considering you don't just easily switch presets with Input Remapper, as you have to "stop" and "inject" another one, which you can actually see take a good second (and it's also recommended to not touch anything), that might not be entirely possible. - That's basically what "Hypershift" does in the Synapse-software seamlessly, except just within the same preset/profile.
But that's really the key, to be able to just transform the entire layout by holding down any designated key to quickly have other inputs at your fingers. - I've used this for games/programs that take up more than 20-25 keys and buttons my keypad has, particularly games that have a lot of items or tools or many inputs for different modes (like building and management games).

But again, inform me if what you explained is capable of achieving that.

**Reply by jonasBoss** [reply]
> Is that what can be done with what you described?

I think so, yes. It might not be as convenient as a "Layer" or Preset. But it should work the same.
With this (`Shift R + KP End` -> `a`) I mean a mapping from a input to an output (`input`->`output`)
So `a`->`b` would mean that you press `a` on your keyboard but it sends a `b` to the computer. InputRemapper also works with key combinations: so `Alt L + a`->`b` would mean that you press the combination of `left alt` and `a` and inputremapper outputs `b`. Below is a screenshot for clarity.
![Screenshot_20230617_153755](https://github.com/sezanzeb/input-remapper/assets/31804124/8a1aa853-cee1-4808-a17c-7fe28a7faaf9)

So now you can create a bunch of mappings that all use a input-combination of the key you use to switch the layer and the key you want to change in that layer.

 - `layer key + button 1` -> `output 1`
 - `layer key + button 2` -> `output 2`
 - `layer key + button 3` -> `output 3`
 - `layer key + button 4` -> `output 4`

**Reply by ObiWaz** [reply]
I know this is an older thread, but I wanted to reply as I have gone down this rabbit hole. This doesn't really work. Especially in your example with alt. I have a Razer Tartarus Pro. One of the buttons is assigned to Alt L. Unfortunately, it is the most convenient button to assign to a layer shifting key. When I assign Alt L + c -> m for example, most programs just assume I am pressing alt, so the key isn't sufficiently remapped. Some games will recognize it, but if I remap in game it will remap with alt + m (when I press alt + c) 

Is there even a way to remap ctrl or alt keys with this software?

I've been trying to solve this issue for some time now and I just don't think there is a way to do this without Razer's very intrusive software.

**Reply by Tesla-Tank** [reply]
> Is there even a way to remap ctrl or alt keys with this software?

I've been using the `set(foo, 1).hold().set(foo, 0)` as the hypershift button and used this to remap the key I want `if_eq($foo, 1, hold(escape), hold(t))`. The first key `escape` will be used when the hypershift button is active, when its not active `t` would be used instead. I couldn't find the original section for this. but I think this explanation can help: [https://github.com/sezanzeb/input-remapper/blob/main/readme/examples.md#combinations-spanning-multiple-devices](url)
